components/browser/session/src/main/java/mozilla/components/browser/session/storage/DefaultSessionStorage.kt

```diff
@@ -85,41 +87,43 @@ class DefaultSessionStorage(
                     val jsonSession = jsonRoot.getJSONObject(it)
                     val session = deserializeSession(it, jsonSession.getJSONObject(SESSION_KEY))
                     val engineSession = deserializeEngineSession(
-                        sessionManager.engine,
-                        jsonSession.getJSONObject(ENGINE_SESSION_KEY))
-                    sessionManager.add(session, engineSession = engineSession)
-                }
-
-                sessionManager.findSessionById(selectedSessionId)?.let { session ->
-                    sessionManager.select(session)
+                            engine,
+                            jsonSession.getJSONObject(ENGINE_SESSION_KEY))
+
+                    sessions.add(SessionWithState(
+                            session,
+                            session.id == selectedSessionId,
+                            engineSession
+                    ))
                 }
             }
-            true
         } catch (_: IOException) {
-            false
+            return null
         } catch (_: JSONException) {
-            false
+            return null
         }
+
+        return sessions
     }
 
     @Synchronized
-    override fun persist(sessionManager: SessionManager): Boolean {
+    override fun persist(engine: Engine, sessions: SessionsSnapshot): Boolean {
         var file: AtomicFile? = null
         var outputStream: FileOutputStream? = null
 
         return try {
             val json = JSONObject()
             json.put(VERSION_KEY, VERSION)
-            json.put(SELECTED_SESSION_KEY, sessionManager.selectedSession?.id ?: "")
+            json.put(SELECTED_SESSION_KEY, sessions.find { it.selected }?.session?.id ?: "")
```

Comment:
It seems unfortunate that we have to go search for the selected session although that is already known information. Should we just add the selected session Id to the `SessionsSnapshot`? (see @pocmo's comment below.) So, this would be another argument for turning in into a class.